docs/articles_en/documentation/openvino-ir-format/operation-sets/operation-specs/internal/paged-attention.rst

```diff
@@ -0,0 +1,129 @@
+
+PagedAttention
+==============
+
+.. meta::
+   :description: Learn about PagedAttention - an attention operator for speculative decoding with in-kernel cache eviction.
+
+**Versioned name**: *PagedAttention*
+
+**Category**: *Sequence processing*
+
+**Short description**: *PagedAttention* implements an optimized attention operator using a block-based KV cache system to enable speculative decoding, efficient memory reuse, and scalable transformer inference.
+
+---
+
+**Detailed description**:
+
+PagedAttention enables high-performance transformer inference by implementing blockwise key-value (KV) memory layouts, dynamic input chunking, and optional cache eviction logic. It efficiently attends over a combination of past and present tokens across multiple sequences with varying lengths and independently managed memory blocks.
+
+---
+
+**Algorithm Explanation**
+
+At the core, attention is computed as:
+
+\[
+\text{Attention}(Q, K, V) = \text{softmax}\left(\frac{QK^T}{\sqrt{d_k}} + \text{Bias}\right)V
+\]
+
+**With enhancements:**
+
+- **KV Cache Layout**: Keys and values are stored in *blocks* of shape `[block_size, head_size]` indexed by `block_indices`. Each sequence may use different blocks for its context.
+- **Subsequence Tokens**: The `subsequence_begins` input defines token start offsets for each sequence in the token batch.
+- **Block Insertion**: When new tokens are processed, their `K` and `V` are inserted into KV cache blocks at locations determined by `block_indices`, updating these blocks in-place.
+- **Optional Eviction Logic**: Controlled externally or internally (if enabled via kernel logic), older blocks may be replaced to maintain memory limits.
+
+**Note:** `block_indices` does not track which blocks are empty or free. It assumes the external memory manager has already assigned valid, available blocks for insertion.
+
+---
+
+**KV Cache Insertion – Pseudocode**
+
+.. code-block:: python
+
+    def insert_into_cache(key_cache, value_cache, key, value, block_indices, block_starts, block_size):
+        """
+        Insert new key/value tensors into cache blocks.
+        Assumes all block_indices have already been allocated and are valid.
+        """
+        token_idx = 0
+        for seq_id, block_begin in enumerate(block_starts[:-1]):
+            seq_token_count = block_starts[seq_id + 1] - block_begin
+            # Blocks assigned to this sequence
+            assigned_blocks = block_indices[block_begin : block_begin + (seq_token_count + block_size - 1) // block_size]
+
+            for i in range(seq_token_count):
+                block_id = assigned_blocks[i // block_size]
+                offset = i % block_size
+                key_cache[block_id, :, offset, :] = key[token_idx]
+                value_cache[block_id, :, offset, :] = value[token_idx]
+                token_idx += 1
+
+---
+
+**Inputs**
+
+1. **query** – `[batch_size_in_tokens, num_heads * head_size]`, type *T*
```

Comment:
It should be better to explain `batch_size_in_tokens`, `batch_size_in_sequences` etc.